dulwich/porcelain.py

```diff
@@ -1239,28 +1274,35 @@ def pull(
     errstream=default_bytes_err_stream,
     fast_forward=True,
     force=False,
-    **kwargs,
+    refspec_encoding=DEFAULT_ENCODING,
+    filter_spec=None,
+    **kwargs
 ):
     """Pull from remote via dulwich.client.
 
     Args:
       repo: Path to repository
       remote_location: Location of the remote
-      refspecs: refspecs to fetch
+      refspecs: refspecs to fetch. Can be a bytestring, a string, or a list of
+        bytestring/string.
       outstream: A stream file to write to output
       errstream: A stream file to write to errors
+      refspec_encoding: Character encoding of bytestrings provided in the refspecs parameter.
+        If not specified, the internal default encoding will be used.
+      filter_spec: A git-rev-list-style object filter spec, as an ASCII string.
```

Comment:
> What does C Git do if --filter is specified and the remote server doesn't support filtering?

The client has no way to tell whether the server will do any filtering. The first message sent by the server already includes the list of references, and the decision whether to filter this list happens server-side.

This parameter is more of a hint for the server, to avoid unnecessary work when the client doesn't need the full list of references. However, the Git docs explicitly state that the server is free to ignore this parameter and just provide all the refs anyway.

Reply:
ack, thanks!

Reply:
Actually you may have been asking about the object filter rather than the reference filter?

A lack of the object filtering capability could be treated as an error if the user has requested object filtering.

However, Git does not treat this as a fatal error, it prints a warning and proceeds with a full clone:

Cloning into bare repository 'test-clone'...                                                   
warning: filtering not recognized by server, ignoring

I am not sure if dulwich is supposed to print such a warning?

I suppose we could provide some way for API users to detect the lack of server-side object filtering, e..g. via a callback, and let them decide whether the transfer should be aborted early. I would keep such an enhancement for later though. For now, matching Git's fallback behaviour seems good enough to me.